### ISSUE TITLE
Middleman fails to start if local hostname includes a space

### ISSUE THREAD

**Issue by johanhalse** [middleman/middleman#1510]
I'm on a mac. Using Ruby 2.2.2 and Middleman 3.3.12. Starting the server from CLI gives the following error:
```
/Users/johanhalse/.rbenv/versions/2.2.0/lib/ruby/2.2.0/uri/rfc3986_parser.rb:66:in `split': bad URI(is not URI?): http://Halse Pro:4567/ (URI::InvalidURIError)
```

Since there's a space in the hostname the URI needs to be encoded to work. I fixed it locally by replacing [line 267 in preview_server.rb](https://github.com/middleman/middleman/blob/v3-stable/middleman-core/lib/middleman-core/preview_server.rb#L267) with
```ruby
def uri(host = URI.encode(Socket.gethostname))
```
I'd submit a pull request but I couldn't find the tests for the preview server :)

**Comment by tdreyno** [comment]
Interesting. Ping @bhollis

**Comment by Arcovion** [comment]
It is indeed a bad URI, since [hostnames can't have spaces](http://en.wikipedia.org/wiki/Hostname#Restrictions_on_valid_host_names)...
But the simplest workaround is: `Socket.gethostname.tr(' ', '+')`
`URI::encode`/`URI::escape` is deprecated by the way, in favour of `CGI::escape`

**Comment by Arcovion** [comment]
@johanhalse please try:
```ruby
gem 'middleman', github: 'middleman', branch: 'hostname'
```

**Comment by bhollis** [comment]
Wow, I've never seen a hostname with spaces.

**Comment by johanhalse** [comment]
@Arcovion @bhollis I'm not even sure how it happened, if I set it myself a few years ago or if it has to do with my computer's sharing name having emoji in it. But there it is! Thanks for your help :bow:

**Comment by dg-ratiodata** [comment]
http://tools.ietf.org/html/rfc952: No blank or space characters are permitted as part of a  name.

So it's not really a bug, that `middleman` cannot handle white space in the hostname. I think the only thing missing here is a check for validity of hostname. If the name is invalid use an ip address.

**Comment by eric-hu** [comment]
I ran into the same issue when upgrading Middleman from an older version.  I had trouble finding where in OS X the name is actually configured.  Here's what I found:

- the computer name that `Socket.gethostname` reports isn't available in the OS X system settings
- instead, it matched the result of `scutil --get HostName`
- I was able to remove a space from my computer name by using `scutil --set HostName "Erics-laptop"`

**Comment by tdreyno** [comment]
Should we re-open this?

**Comment by tdreyno** [comment]
cc @Arcovion @dg-ratiodata

**Comment by dg-ratiodata** [comment]
@eric-hu Which version are you using? From git or default rubygems.

@tdreyno Added some code to the new implementation to remove "white space" in hostname. Did you release the code in v3-stable as new version on rubygems.org?

**Comment by eric-hu** [comment]
@dg-ratiodata this was only using Rubygems.  I changed my Gemfile like so:

    -gem "middleman", "~> 3.1.4"
    +gem "middleman", "~> 3.1"

And with that, ran `bundle update`.  My Gemfile.lock now shows `middleman (3.3.12)`.

FYI, I was able to resolve my issue by simply removing the whitespace from my computer name.  Your response back in May about not having whitespace in a network node name seems reasonable to me.  I commented because it was non-obvious how to change this value in OS X.
